cpp/src/column/column_device_view.cu

```diff
@@ -32,7 +32,6 @@ column_device_view::column_device_view(column_view source)
 
 // Free device memory allocated for children
 void column_device_view::destroy() {
-  // TODO Needs to handle grand-children
   RMM_FREE(d_children,0);
   delete this;
```

Comment:
This is required because we are overriding the default deleter for `column_device_view` as per the design of `std::unique_ptr` and must therefore do the delete ourselves. The override is needed to free the device memory held by the class instance here:
https://github.com/rapidsai/cudf/blob/2778adb2efd50658193f2c8d8697169d7fe969da/cpp/src/column/column_device_view.cu#L36
https://isocpp.org/wiki/faq/freestore-mgmt#delete-this

The https://github.com/rapidsai/cudf/blob/51c9e0e5d5ab1a2ab458abca4cbf0b500cbc0b38/cpp/include/cudf/column/column_device_view.cuh#L344 line says
_"Does not free the column data, simply free's the device memory allocated to hold the child views."_
This statement is correct. The `column_device_view` points to device memory held by a `column` class. This memory is not touched when the view is destroyed. There is some device memory to hold the child views that is unique to the `column_device_view` and not pointed to by the original `column` class.

Reply:
Thank you very much for the explanation, and I still have one unresolved question.

This destroy() function is a public member function implying that this function can be called elsewhere (besides the custom deleter).  And the user may not know this calls "delete this" unless one digs into the code.

```
1. You must be absolutely 100% positively sure that this object was allocated via new (not by new[], nor by placement new, nor a local object on the stack, nor a namespace-scope / global, nor a member of another object; but by plain ordinary new).
2. You must be absolutely 100% positively sure that your member function will be the last member function invoked on this object.
3. You must be absolutely 100% positively sure that the rest of your member function (after the delete this line) doesn’t touch any piece of this object (including calling any other member functions or touching any data members). This includes code that will run in destructors for any objects allocated on the stack that are still alive.
4. You must be absolutely 100% positively sure that no one even touches the this pointer itself after the delete this line. In other words, you must not examine it, compare it with another pointer, compare it with nullptr, print it, cast it, do anything with it.
```
https://isocpp.org/wiki/faq/freestore-mgmt#delete-this

Doesn't this increase the risk of one of the above conditions getting violated?

Reply:
The `destroy()` needs to be public because it is called automatically by the `std::unique_ptr`. The same rule applies for the destructor for the default deleter.

Reply:
The doxygen comment for destroy:
https://github.com/rapidsai/cudf/blob/51c9e0e5d5ab1a2ab458abca4cbf0b500cbc0b38/cpp/include/cudf/column/column_device_view.cuh#L342

Reply:
Just noticed that name typo :-)

Reply:
I guess I'm missing something, but can't we just call RMM_FREE(d_children) inside the deleter of column_device_view? (instead of using the default destructor (~column_device_view() = default) which will not free d_children). If I'm not mistaken, then, neither calling "delete this" inside the public destroy function nor custom deleter will be necessary.

Reply:
And if doing this is unavoidable, I think it is safer to at least document that this function is not for public use (but public for an unavoidable reason) in https://github.com/rapidsai/cudf/blob/51c9e0e5d5ab1a2ab458abca4cbf0b500cbc0b38/cpp/include/cudf/column/column_device_view.cuh#L342

Reply:
The deleter is defined here: https://github.com/rapidsai/cudf/blob/61278219a0b4ffe9f6054db303d22d90f300084c/cpp/src/column/column_device_view.cu#L71
The `RMM_FREE` cannot be placed here directly because `d_children` is protected member.
And we cannot call `RMM_FREE` inside the destructor because the destructor may be called inside device code. An instance of this class (or any of our *_device classes) easily passed between host and device code.
The `destroy()` method is called only when the `std::unique_ptr` is done with the object (which only happens in CPU code).

If you feel this is an that this should be changed or documented better, I think it best to open a separate github issue since I feel this issue is outside the scope of what is intended by this PR.

Reply:
Ah... OK, I wasn't aware that "we cannot call RMM_FREE inside the destructor because the destructor may be called inside device code." Makes sense, and thanks for the clarification.